QuadradiusR/scripts/ws_api.gd

```diff
@@ -17,24 +31,80 @@ func _ready():
 func _process(delta):
 	ws.poll()
 
+
 # # # # # # # # # #
 # Signal Handlers #
 # # # # # # # # # #
 func _on_connection_closed(was_clean: bool):
 	print('connection closed, clean: ', was_clean)
+	set_process(false)
 
 func _on_connection_error():
 	print('connection error')
+	set_process(false)
 
 func _on_connection_established(protocol: String):
-	print('im in')
-	ws.get_peer(1).put_packet(JSON.print({'op': 2, 'd': {"token": token}}).to_utf8())
+	print('im in')  # DEBUG
+	_send_data({'op': 2, 'd': {"token": token}})  # authorize with the server before doing anything else
 
 func _on_data_received():
-	print('data received:', ws.get_peer(1).get_packet().get_string_from_utf8())
+	var data = _get_data()
+	print('data received:', data)  # DEBUG
+	op_handlers[int(data['op'])].call_func(data)
```

Comment:
JSON has `int` type, this casting  `int(data['op'])` is needed? `op_handlers[data['op']]` is a bit easier to read 😸